plugin/trino-hive/src/test/java/io/trino/plugin/hive/fs/TestCachingDirectoryListerRecursiveFilesOnly.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.hive.fs;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.airlift.units.Duration;
+import io.trino.plugin.hive.metastore.MetastoreUtil;
+import io.trino.plugin.hive.metastore.Table;
+import io.trino.testing.QueryRunner;
+import org.apache.hadoop.fs.Path;
+import org.testng.annotations.Test;
+
+import java.util.List;
+import java.util.NoSuchElementException;
+
+import static io.trino.plugin.hive.HiveQueryRunner.TPCH_SCHEMA;
+import static io.trino.plugin.hive.metastore.PrincipalPrivileges.NO_PRIVILEGES;
+import static java.lang.String.format;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertTrue;
+
+// some tests may invalidate the whole cache affecting therefore other concurrent tests
+@Test(singleThreaded = true)
+public class TestCachingDirectoryListerRecursiveFilesOnly
+        extends BaseCachingDirectoryListerTest<CachingDirectoryLister>
+{
+    @Override
+    protected CachingDirectoryLister createDirectoryLister()
+    {
+        return new CachingDirectoryLister(Duration.valueOf("5m"), 1_000_000L, List.of("tpch.*"));
+    }
+
+    @Override
+    protected QueryRunner createQueryRunner()
+            throws Exception
+    {
+        return createQueryRunner(ImmutableMap.of(
+                "hive.allow-register-partition-procedure", "true",
+                "hive.recursive-directories", "true"));
+    }
+
+    @Override
+    protected boolean isCached(CachingDirectoryLister directoryLister, Path path)
+    {
+        return directoryLister.isCached(new DirectoryListingCacheKey(path, true));
+    }
+
+    @Test
+    public void testRecursiveDirectories()
+    {
+        // Create partitioned table to force files to be inserted in sub-partition paths
+        assertUpdate("CREATE TABLE recursive_directories (clicks bigint, day date, country varchar) WITH (format = 'ORC', partitioned_by = ARRAY['day', 'country'])");
+        assertUpdate("INSERT INTO recursive_directories VALUES (1000, DATE '2022-02-01', 'US'), (2000, DATE '2022-02-01', 'US'), (4000, DATE '2022-02-02', 'US'), (1500, DATE '2022-02-01', 'AT'), (2500, DATE '2022-02-02', 'AT')", 5);
+
+        // Replace the partitioned table a new unpartitioned table with the same root location, leaving the data in place
+        Table partitionedTable = getTable(TPCH_SCHEMA, "recursive_directories")
```

Comment:
Instead of doing the low-level table dropping and creation, could we create two tables with the same `external_location`:
- for writing use the partitioned configuration `partitioned_by = ARRAY['day', 'country']`
- for reading use the plain configuration without partioning on which we can use the list recursively files super power

This seems to me (although it screams like an anti-pattern) the way in which such a scenario could be implemented in the real-world.

Please take this note with a grain of salt.

Reply:
I tried to do something along those lines, but I couldn't get it work very easily in the way the tests are set up- so I just went with this approach for the purposes of this test. It's not intended to represent a realistic usage pattern outside of this test scenario.